docs/source/csdl-spec.md

```diff
@@ -0,0 +1,630 @@
+
+### Portability
+A type is *portable to* a subgraph if it has a `@csdl_key` for that subgraph.
+
+### Free / Bound fields
+
+Free fields can be resolved by any subgraph, bound fields can only be resolved by the subgraph they're bound to.
+
+## Directives
+### `@csdl`
+<code class='grammar'>
+directive @csdl(version: csdl_Version!) on SCHEMA
+</code>
+
+Specify the version of CSDL needed by this document.
+
+<a class="listing">Using `@csdl` to specify the CSDL version for the document</a>
+
+```graphql
+schema @csdl(version: "^1") {
+  query: Query
+}
+```
+
+### `@csdl_key`
+<code class="grammar">
+directive @csdl_key(graph: csdl_Graph!, repr: csdl_SelectionSet!)
+  repeatable on OBJECT
+</code>
+
+Define an entity key for this type within a subgraph.
+
+The `@csdl_key` directive tells consumers what subset of fields are necessary to identify this type of entity to a particular subgraph. It provides a way for csdl consumers to "switch graphs" when planning a query. For example:
+
+<a name=query-port-astronaut-type href=#query-port-astronaut-type class=listing>A query which requires porting the Astronaut type between services</a>
+```graphql
+query {          # resolver                    | realizes
+                 # ----------------------------|--------------
+  rockets {      # rockets::Query.rockets     -> rockets::Rocket
+    captain {    # rockets::Rocket.astronaut  -> rockets::Astronaut
+      name       # astronauts::Astronaut.name -> String
+      rocket {   # rockets::Astronaut.rocket  -> rockets::Rocket
+        id       # rockets::Rocket.id         -> ID
+      }
+    }
+  }
+  
+}
+```
+
+`Astronaut.name` is provided by the `astronauts` subgraph. But the `captain` field is provided by the `rockets` subgraph.
+
+The fields specified in `repr` will be passed to the subgraph's `Query._entities(representations:)` as an item within the `representations` list.
+
+Multiple `@csdl_key`s can be provided for different graphs, or for the same graph.
+
+<a name=using-csdl_key href=#using-csdl_key class=listing>Using `@csdl_key` to specify subgraph keys</a>
+
+```graphql
+type Astronaut
+  @csdl_key(graph: ASTRONAUTS, repr: "{ id }")
+  @csdl_key(graph: ROCKETS, repr: "{ id }")
+{ ... }
+```
+
+### `@csdl_endpoint`
+<code class='grammar'>
+directive @csdl_endpoint(url: csdl_Url) on ENUM_VALUE
+</code>
+
+Bind an endpoint URL to a subgraph. This directive is only valid on enum values within the required `csdl_Graph` enum type.
+
+<a name=using-csdl_endpoint href=#using-csdl_endpoint class="listing">Using `@csdl_endpoint` to specify subgraph endpoints</a>
+
+```graphql
+enum csdl_Graph {
+  ROCKETS @csdl_endpoint(url: "https://rockets.api.com"),
+  ASTRONAUTS @csdl_endpoint(url: "https://astronauts.api.com"),
+}
+```
+
+### `@csdl_resolve`
+<code class='grammar'>
+directive @csdl_resolve(
+  graph: csdl_Graph!,
+  requires: csdl_SelectionSet,
```

Comment:
Is it required that the selection set listed here is a valid selection set of the parent type? Would we say that's a convenience function, but in theory the planner could still support a root query in this place? (Somewhat prodding in a direction that might be enabling and somewhat asking about the reality of the world)

Reply:
we're thinking along the same lines! I think in a future version of the spec we'll probably introduce a `csdl_FnQuery` input and more general `@csdl_resolve(query:)` form for this.

for the moment, i'd like to ship a spec version that we can support with only very minimal query planner changes.